index.js

```diff
@@ -1,22 +1,39 @@
 const waterfall = require('run-waterfall')
+const pullAsyncMap = require('pull-stream/throughs/async-map')
 
 module.exports = cliPrompter
 
 const promptTypes = require('./types')
 
 function cliPrompter (questions, cb) {
+  if (arguments.length === 0) {
+    return pullAyncScan({}, (values, question, cb) => {
+      const prompter = getPrompter(question)
+      prompter({ question, values }, cb)
+    })
+  }
+
   waterfall([
     cb => cb(null, {}), // start the waterfall with values
-    ...questions.map(promptQuestion)
+    ...questions.map(question => {
+      const prompter = getPrompter(question)
+      return (values, cb) => {
+        prompter({ question, values }, cb)
+      }
+    })
   ], cb)
 }
 
-function promptQuestion (question) {
-  const { type } = question
-  const Prompter = promptTypes[type]
-
-  return (values, cb) => {
-    Prompter({ question, values }, cb)
-  }
+function getPrompter ({ type }) {
+  return promptTypes[type]
 }
 
+function pullAyncScan (value, accumulator) {
+  return pullAsyncMap((next, cb) => {
+    accumulator(value, next, (err, nextValue) => {
+      if (err) return cb(err)
+      value = nextValue
+      cb(null, value[Object.keys(value)[0]])
```

Comment:
@ahdinosaur Ooh.  That makes sense now!  I didn't realize they accumulated like that.

I agree with all those points at the end.  They help make an app more dynamic and a better experience for the user.  We accomplish the same thing in different ways.

The way I was going to build it is: you have a source stream of question objects, and the prompt module maps those to what the user inputs.

The sink of the stream there pretty much works as an emitter for answers.  Use different style of sinks for different purposes, notably `pull.drain` and `pull.sink`.

With `pull.drain`, you can get them as they are answered, update your state (using `redux`/`pull-stream-model` for example), and ask the next question based on the new state.

```js
// Listen for questions (w/ pull-notify), and prompt the user:
pull(
  question.listen(),
  prompter(),
  drain(answer => {
    // update state:
    app.dispatch({ type: 'answer', answer })
    // ask next question:
    var your_cats_name = app.getState().your_cats_name
    question({
      type: 'input',
      message: `is ${your_cats_name} hungry?`
    })
  })
)
```

Another way you can do prompts with `pull.collect`: you can just map a bunch of questions to answers, without any need for state.

```js
// Just map some questions to answers:
pull(
  pull.values([
    { type: 'input', message: 'what is name of your pet unicorn?' },
    { type: 'confirm', message: 'is cat jealous of new unicorn?' },
    // ...
  ]),
  prompt(),
  pull.collect(function (err, answers) {
    // ...
  })
)
```

There might be other interesting systems you can do with that using through streams to do more "level" analysis of the input. I like the open-endedness, while it is also still straightforward I think.

What do you think? I think it has the same values, but they accumulate in the sink instead of in the object